x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/ZeroShotClassificationProcessor.java

```diff
@@ -0,0 +1,195 @@
+        }
+        return new ResultProcessor(entailmentPos, contraPos, labels, isMultiLabel);
+    }
+
+    static class RequestBuilder implements NlpTask.RequestBuilder {
+
+        private final NlpTokenizer tokenizer;
+        private final String[] labels;
+        private final String hypothesisTemplate;
+
+        RequestBuilder(NlpTokenizer tokenizer, String[] labels, String hypothesisTemplate) {
+            this.tokenizer = tokenizer;
+            this.labels = labels;
+            this.hypothesisTemplate = hypothesisTemplate;
+        }
+
+        @Override
+        public NlpTask.Request buildRequest(List<String> inputs, String requestId) throws IOException {
+            if (inputs.size() > 1) {
+                throw new IllegalArgumentException("Unable to do zero-shot classification on more than one text input at a time");
+            }
+            List<TokenizationResult.Tokenization> tokenizations = new ArrayList<>(labels.length);
+            for (String label : labels) {
+                tokenizations.add(tokenizer.tokenize(inputs.get(0), LoggerMessageFormat.format(null, hypothesisTemplate, label)));
+            }
+            TokenizationResult result = tokenizer.buildTokenizationResult(tokenizations);
+            return buildRequest(result, requestId);
+        }
+
+        @Override
+        public NlpTask.Request buildRequest(TokenizationResult tokenizationResult, String requestId) throws IOException {
+            return tokenizer.requestBuilder().buildRequest(tokenizationResult, requestId);
+        }
+    }
+
+    static class ResultProcessor implements NlpTask.ResultProcessor {
+        private final int entailmentPos;
+        private final int contraPos;
+        private final String[] labels;
+        private final boolean isMultiLabel;
+
+        ResultProcessor(int entailmentPos, int contraPos, String[] labels, boolean isMultiLabel) {
+            this.entailmentPos = entailmentPos;
+            this.contraPos = contraPos;
+            this.labels = labels;
+            this.isMultiLabel = isMultiLabel;
+        }
+
+        @Override
+        public InferenceResults processResult(TokenizationResult tokenization, PyTorchResult pyTorchResult) {
+            if (pyTorchResult.getInferenceResult().length < 1) {
+                return new WarningInferenceResults("Zero shot classification result has no data");
+            }
+            // TODO only the first entry in the batch result is verified and
+            // checked. Implement for all in batch
+            if (pyTorchResult.getInferenceResult()[0].length != labels.length) {
+                return new WarningInferenceResults(
+                    "Expected exactly [{}] values in zero shot classification result; got [{}]",
+                    labels.length,
+                    pyTorchResult.getInferenceResult().length
+                );
+            }
+            final double[] normalizedScores;
+            if (isMultiLabel) {
```

Comment:
`isMultiLabel` is just about how the results are interpreted? When true the probability of entailment as opposed to contradiction for each label is returned. When false it is the probability of each label being entailment

Reply:
`isMultiLabel` is basically:

 - When `true` its `softmax` of individual entailment vs contradiction (probs don't sum to 1.0)
 - When `false` its `softmax` of all entailments (probs sum to 1.0)

The docs already state you use it when you could have more than one true label. Which is exactly what we use it for.

I would rather not talk about softmax, entailment, etc.